numpy/__init__.pyi

```diff
@@ -1231,3 +1231,113 @@ def amin(
     initial: _NumberLike = ...,
     where: _ArrayLikeBool = ...,
 ) -> Union[number, ndarray]: ...
+
+# TODO: `np.prod()``: For object arrays `initial` does not necasarily
+# have to be a numerical scalar.
+# The only requirement is that it is compatible
+# with the `.__mult__()` method(s) of the passed array's elements.
+
+# Note that the same holds for `np.sum()` and `.__add__()`.
+
+@overload
+def prod(
+    a: _Number,
+    axis: Optional[_ShapeLike] = ...,
+    dtype: DtypeLike = ...,
+    out: Optional[ndarray] = ...,
+    keepdims: bool = ...,
+    initial: _NumberLike = ...,
+    where: _ArrayLikeBool = ...,
+) -> _Number: ...
+@overload
+def prod(
+    a: ArrayLike,
+    axis: None = ...,
+    dtype: DtypeLike = ...,
+    out: Optional[ndarray] = ...,
+    keepdims: Literal[False] = ...,
+    initial: _NumberLike = ...,
+    where: _ArrayLikeBool = ...,
+) -> number: ...
+@overload
+def prod(
+    a: ArrayLike,
+    axis: Optional[_ShapeLike] = ...,
+    dtype: DtypeLike = ...,
+    out: Optional[ndarray] = ...,
+    keepdims: bool = ...,
+    initial: _NumberLike = ...,
+    where: _ArrayLikeBool = ...,
+) -> Union[number, ndarray]: ...
+def cumprod(
+    a: ArrayLike,
+    axis: Optional[int] = ...,
+    dtype: DtypeLike = ...,
+    out: Optional[ndarray] = ...,
+) -> ndarray: ...
+def ndim(a: ArrayLike) -> int: ...
+def size(a: ArrayLike, axis: Optional[int] = ...) -> int: ...
+@overload
+def around(
+    a: _Number, decimals: int = ..., out: Optional[ndarray] = ...
+) -> _Number: ...
+@overload
+def around(
+    a: _NumberLike, decimals: int = ..., out: Optional[ndarray] = ...
+) -> number: ...
+@overload
+def around(
+    a: ArrayLike, decimals: int = ..., out: Optional[ndarray] = ...
+) -> ndarray: ...
+@overload
+def mean(
+    a: ArrayLike,
+    axis: None = ...,
+    dtype: DtypeLike = ...,
+    out: Optional[ndarray] = ...,
+    keepdims: Literal[False] = ...,
+) -> number: ...
+@overload
+def mean(
+    a: ArrayLike,
+    axis: Optional[_ShapeLike] = ...,
+    dtype: DtypeLike = ...,
+    out: Optional[ndarray] = ...,
+    keepdims: bool = ...,
+) -> Union[number, ndarray]: ...
+@overload
+def std(
+    a: ArrayLike,
+    axis: None = ...,
+    dtype: DtypeLike = ...,
+    out: Optional[ndarray] = ...,
+    ddof: int = ...,
+    keepdims: Literal[False] = ...,
+) -> number: ...
+@overload
+def std(
+    a: ArrayLike,
+    axis: Optional[_ShapeLike] = ...,
+    dtype: DtypeLike = ...,
+    out: Optional[ndarray] = ...,
+    ddof: int = ...,
+    keepdims: bool = ...,
+) -> Union[number, ndarray]: ...
+@overload
+def var(
+    a: ArrayLike,
+    axis: None = ...,
+    dtype: DtypeLike = ...,
+    out: Optional[ndarray] = ...,
+    ddof: int = ...,
+    keepdims: Literal[False] = ...,
+) -> number: ...
```

Comment:
`number` isn't correct when `out` is specified:
```
>>> np.var([1], out=np.zeros(()))
array(0.)
```

Reply:
This applies throughout.

Reply:
> This applies for most but not all functions in this patch with `out`.

Considering the widespread nature of this issue I'd propose to save it for a future maintenance pull request and fix it all at once.

Reply:
In the meantime then, it would probably be a good idea to annotate these with `Union[number, ndarray]` - weak annotations are better than incorrect ones.

Reply:
As it turns out is fairly easy way to fix this: `out: Optional[ndarray]` must be changed into `out: None` when a `number` is returned.

I have noticed some inconsistent behavior when using `out` in combination with passing a `number` though,
as in such case another `number` is returned rather than a `ndarray`.
Builtin scalars, builtin sequences and `ndarray`s do all return an `ndarray` (see example below).

``` python
In [1]: import numpy as np                                                                                                  

In [2]: np.__version__                                                                                                      
Out[2]: '1.20.0.dev0+62f26cf'

In [3]: np.var(np.int64(1), out=np.zeros(()))                                                                               
Out[3]: 0.0

In [4]: np.var(1, out=np.zeros(()))                                                                                         
Out[4]: array(0.)

In [5]: np.var([1], out=np.zeros(()))                                                                                       
Out[5]: array(0.)

In [6]: np.var(np.ones(1), out=np.zeros(()))                                                                                
Out[6]: array(0.)
```

Reply:
That is arguably a bug in `np.var`, when `out` is given, it must be the return value.

Reply:
> `number` isn't correct when `out` is specified:

Fixed as of https://github.com/numpy/numpy/pull/16729/commits/5a4244029794a4835469d1574d472f99a7f55f4b.

Reply:
> That is arguably a bug in `np.var`, when `out` is given, it must be the return value.

I'll double check which (currently annotated) functions are affected by this this and create an issue afterwards.